cluster.go

```diff
@@ -0,0 +1,316 @@
+	// https://github.com/antirez/redis-rb-cluster/blob/fd931ed/cluster.rb#L157
+	for addr := range c.addrs {
+		c.reset()
+
+		infos, err = c.fetchClusterSlots(addr)
+		if err == nil {
+			c.update(infos)
+			break
+		}
+	}
+	return
+}
+
+// Closes all connections and flushes slots cache
+func (c *ClusterClient) reset() error {
+	err := c.conns.Clear()
+	c.slots = make([][]string, hashSlots)
+	return err
+}
+
+// Forces a cache reload on next request
+func (c *ClusterClient) forceReload() {
+	atomic.StoreUint32(&c._reload, 1)
+}
+
+// Find the next untried address
+func (c *ClusterClient) findNextAddr(tried map[string]struct{}) string {
+	for addr := range c.addrs {
+		if _, ok := tried[addr]; !ok {
+			return addr
+		}
+	}
+	return ""
+}
+
+// Fetch slot information
+func (c *ClusterClient) fetchClusterSlots(addr string) ([]ClusterSlotInfo, error) {
+	client := NewClient(c.opt.ClientOptions(addr))
+	defer client.Close()
+
+	return client.ClusterSlots().Result()
+}
+
+// Update slot information, populate slots
+func (c *ClusterClient) update(infos []ClusterSlotInfo) {
+	for _, info := range infos {
+		for i := info.Start; i <= info.End; i++ {
+			c.slots[i] = info.Addrs
+		}
+
+		for _, addr := range info.Addrs {
+			c.addrs[addr] = struct{}{}
+		}
+	}
+}
+
+//------------------------------------------------------------------------------
+
+var errNoAddrs = errors.New("redis: no addresses")
+
+type ClusterOptions struct {
+	// A seed-list of host:port addresses of known cluster nodes
+	Addrs []string
+
+	// An optional password
+	Password string
+
+	// The maximum number of MOVED/ASK redirects to follow, before
+	// giving up. Default: 16
+	MaxRedirects int
+
+	// The maximum number of open connections to cluster nodes. Default: 10.
+	// WARNING: Each connection maintains its own connection pool. The maximum
+	// possible number of socket connections is therefore MaxConns x PoolSize
+	MaxConns int
+
+	// The maximum number of TCP sockets per connection. Default: 5
+	PoolSize int
+
+	// Timeout settings
+	DialTimeout, ReadTimeout, WriteTimeout, IdleTimeout time.Duration
+}
+
+func (opt *ClusterOptions) getPoolSize() int {
+	if opt.PoolSize < 1 {
+		return 5
+	}
+	return opt.PoolSize
+}
+
+func (opt *ClusterOptions) getMaxConns() int {
+	if opt.MaxConns < 1 {
+		return 10
+	}
+	return opt.PoolSize
+}
+
+func (opt *ClusterOptions) getDialTimeout() time.Duration {
+	if opt.DialTimeout == 0 {
+		return 5 * time.Second
+	}
+	return opt.DialTimeout
+}
+
+func (opt *ClusterOptions) getMaxRedirects() int {
+	if opt.MaxRedirects < 1 {
+		return 16
+	}
+	return opt.MaxRedirects
+}
+
+func (opt *ClusterOptions) getAddrSet() (map[string]struct{}, error) {
+	size := len(opt.Addrs)
+	if size < 1 {
+		return nil, errNoAddrs
+	}
+
+	addrs := make(map[string]struct{}, size)
+	for _, addr := range opt.Addrs {
+		addrs[addr] = struct{}{}
+	}
+	return addrs, nil
+}
+
+func (opt *ClusterOptions) ClientOptions(addr string) *Options {
```

Comment:
I would prefer we unexport this. But if you think that it must be exported can you please remove `addr` argument from this func? Instead you can always do:

```
clopt = opt.ClientOptions()
clopt.Addr = theAddr
```

This way `ClientOptions() *Options` will have the same signature on `ClusterOptions` and `FailoverOptions`.